worldedit-core/src/main/java/com/fastasyncworldedit/core/queue/implementation/ApplyTask.java

```diff
@@ -0,0 +1,241 @@
+            }
+        } else {
+            // we reached a task for a single chunk, let's process it
+            processChunk(this.minChunkX, this.minChunkZ);
+        }
+        if (this.shift == INITIAL_REGION_SHIFT) {
+            onCompletion();
+        }
+    }
+
+    private void processRegion(int regionX, int regionZ, int shift) {
+        final ThreadState<F> state = getState();
+        this.commonState.parallelQueueExtent.enter(state.queue);
+        try {
+            for (int chunkX = regionX << shift; chunkX <= ((regionX  + 1) << shift) - 1; chunkX++) {
+                for (int chunkZ = regionZ << shift; chunkZ <= ((regionZ  + 1) << shift) - 1; chunkZ++) {
+                    if (!this.commonState.region.containsChunk(chunkX, chunkZ)) {
+                        continue; // chunks not intersecting with the region must not be processed
+                    }
+                    applyChunk(chunkX, chunkZ, state);
+                }
+            }
+        } finally {
+            this.commonState.parallelQueueExtent.exit();
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private ThreadState<F> getState() {
+        return this.commonState.stateCache.computeIfAbsent(
+                Thread.currentThread(),
+                __ -> new ThreadState<>(
+                        (SingleThreadQueueExtent) this.commonState.parallelQueueExtent.getNewQueue(),
+                        (F) this.commonState.originalFilter.fork()
+                )
+        );
+    }
+
+    private void processChunk(int chunkX, int chunkZ) {
+        final ThreadState<F> state = getState();
+        this.commonState.parallelQueueExtent.enter(state.queue);
+        try {
+            applyChunk(chunkX, chunkZ, state);
+        } finally {
+            this.commonState.parallelQueueExtent.exit();
+        }
+    }
+
+    private void applyChunk(int chunkX, int chunkZ, ThreadState<F> state) {
+        try {
+            state.block = state.queue.apply(
+                    state.block,
+                    state.filter,
+                    this.commonState.region,
+                    chunkX,
+                    chunkZ,
+                    this.commonState.full
+            );
+        } catch (Throwable t) {
+            if (t instanceof FaweException faweException) {
+                Fawe.handleFaweException(this.commonState.faweExceptionReasonsUsed, faweException, LOGGER);
+            } else if (t.getCause() instanceof FaweException faweException) {
+                Fawe.handleFaweException(this.commonState.faweExceptionReasonsUsed, faweException, LOGGER);
+            } else {
+                throw t;
+            }
+        }
+    }
+
+    private void onCompletion() {
+        for (ForkJoinTask<?> task : postProcess()) {
+            if (task.tryUnfork()) {
+                task.invoke();
+            } else {
+                task.join();
+            }
+        }
+    }
+
+    private ForkJoinTask<?>[] postProcess() {
```

Comment:
Is postProcess necessarily the best name here? Given the use of postprocess already to define actions per-chunk. Perhaps just flushQueues is fine as we're not doing any processing (postly or otherwise) here